File: programmers/lv3/가장 먼 노드.py
# 인접 리스트
def solution(n, edge):
    graph = [[] for _ in range(n + 1)]
    for _ in edge:
        a, b = _
        graph[a].append(b)
        graph[b].append(a)
    from collections import deque
    q = deque()
    q.append((1, -1))
    visited = [0] * (n + 1)
    visited[1] = 1
    
    result = [0] * (n + 1)
    while q:
        now, cnt = q.popleft()
        result[now] += cnt + 1
        for i in graph[now]:
            if not visited[i]:
                visited[i] = 1
                q.append((i, cnt + 1))
    return result.count(max(result))
        
# 인접 행렬 (타임오버)
def solution(n, edge):
    graph = [[0] * (n + 1) for _ in range(n + 1)]
    for _ in edge:
        a, b = _
        graph[a][b] = 1
        graph[b][a] = 1
    
    from collections import deque
    q = deque()
    q.append((1, -1))
    visited = [0] * (n + 1)
    visited[1] = 1
    
    result = [0] * (n + 1)
    while q:
        now, cnt = q.popleft()
        result[now] += cnt + 1
        for i in range(len(graph[now])):
            if graph[now][i] and not visited[i]:
                visited[i] = 1
                q.append((i, cnt + 1))
    return result.count(max(result))
        

File: programmers/lv3/가장 먼 노드.js
function solution(n, edge) {
  const graph = Array.from({ length: n + 1 }, () => []);
  edge.forEach((e) => {
    const [a, b] = e;
    graph[a].push(b);
    graph[b].push(a);
  });
  const q = [[1, -1]];
  const visited = Array(n + 1).fill(0);
  visited[1] = 1;
  const result = Array(n + 1).fill(0);
  while (q.length) {
    const [now, cnt] = q.shift();
    result[now] += cnt + 1;
    for (const i of graph[now]) {
      if (!visited[i]) {
        visited[i] = 1;
        q.push([i, cnt + 1]);
      }
    }
  }
  const m = Math.max(...result);
  let cnt = 0;
  result.forEach((v) => {
    if (v === m) cnt += 1;
  });
  return cnt;
}
